### DISCUSSION TITLE
Advanced Foudry's DSCEngine design question

### DISCUSSION THREAD

**Discussion by pierrealex3** [Cyfrin/foundry-full-course-cu#2618] [Q&A] [answered]
In the DSCEngine SC, we do maintain a mapping for user to amount of DSC minted in the storage variable s_DSCMinted of the contract.

```
    mapping(address user => uint256 amountDscMinted) private s_DSCMinted;
    DecentralizedStableCoin private immutable i_dsc;
```

Why do we maintain this structure instead of actually rely on a call to our i_dsc.balanceOf(user) to get that info ?
Isn't that dangerous to keep a "potentially" unsynchronized copy of the info we already have in the DecentralizedStableCoin contract?

**Comment by EngrPips** [comment, marked as answer]
Well, It is dangerous to be getting the balance of users on the `dsc` token itself inside of our own protocol. Put it at the back of your mind that internal accounting is always preferred over external accounting. We want to use the record we keep ourselves. What if a user minted and then transferred their token to another user?

**Reply by pierrealex3** [reply]
Thanks for the reply! I am trying to understand the actual danger from the example you said, but I can't figure it out on the top of my mind.  I will probably understand this "internal accounting preferred over external accounting" advice later on in my progress.  I guess it will make sense, as atm my Web2 backend developer brain tells me "double bookkeeping" of the same data -> WRONG 😄

**Reply by EngrPips** [reply]
I understand if it doesn't totally make sense to you right now, but it will surely make more sense over time.
